providers/fips/fipsprov.c

```diff
@@ -1238,3 +1259,128 @@ void OSSL_INDICATOR_get_callback(OSSL_LIB_CTX *libctx,
             *cb = NULL;
     }
 }
+
+/* On Deamnd test infrastructure */
+
+struct ondemand_test_status {
+    const char *alg;
+    int pass;
+};
+
+static struct ondemand_test_status ondemand_kat_integrity_tests[] = {
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_cipher_tests[] = {
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_asym_cipher_tests[] = {
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_asym_keygen_tests[] = {
+    { "SLH-DSA-SHA2-128f", 0 },
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_kem_tests[] = {
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_digest_tests[] = {
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_signature_tests[] = {
+    { "SLH-DSA-SHAKE-128f", 0 },
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_kdf_tests[] = {
+    { NULL, 0 }
+};
+static struct ondemand_test_status ondemand_kat_ka_tests[] = {
+    { NULL, 0 }
+};
+
+static struct ondemand_test_status *ondemand_tests_status[OSSL_ONDEMAND_MAX] = {
+    ondemand_kat_integrity_tests,
+    ondemand_kat_cipher_tests,
+    ondemand_kat_asym_cipher_tests,
+    ondemand_kat_asym_keygen_tests,
+    ondemand_kat_kem_tests,
+    ondemand_kat_digest_tests,
+    ondemand_kat_signature_tests,
+    ondemand_kat_kdf_tests,
+    ondemand_kat_ka_tests,
+};
+
+int FIPS_kat_on_demand(OSSL_LIB_CTX *libctx, int type, const char *alg)
+{
+    FIPS_GLOBAL *fgbl;
+    OSSL_SELF_TEST *ev = NULL;
+    void *local_key;
+    int idx = 0;
+    int ret = 0;
+
+    if ((fgbl = ossl_lib_ctx_get_data(libctx, OSSL_LIB_CTX_FIPS_PROV_INDEX)) == NULL)
+        return 0;
+
+    if (fgbl->ondemand_lock == NULL)
+        return 0;
+
+    if (type > OSSL_ONDEMAND_MAX)
+        return 0;
+
+    for (; ondemand_tests_status[type][idx].alg != NULL; idx++) {
+        if (strcmp(ondemand_tests_status[type][idx].alg, alg) == 0)
+            break;
+    }
+    if (ondemand_tests_status[type][idx].alg == NULL)
+        return 0;
+
+    if (ondemand_tests_status[type][idx].pass)
+        return 1;
+
+    /*
+     * now that we found the test, before we do anything else make sure
+     * a local test is not already in progress or we'll deadlock
+     */
+    local_key = CRYPTO_THREAD_get_local(&fgbl->ondemand_key);
+    if (local_key != NULL) {
+        /* self test occurring in this thread, let it continue */
+        return -1;
+    }
+
+    if (CRYPTO_THREAD_write_lock(fgbl->ondemand_lock)) {
```

Comment:
Why have both a thread local indicating that the test is running and a global lock to run the test under?

I don't see how the thread local helps.  If we're running the test, we're running the test and won't be interrupted until it's finished.  If we're not running it, we'll never run it.

Reply:
The tests will call back in to the functions that called the test right away, so we need a way to make those functions go through and not try to run self tests again in a loop while we are testing. But we do not want any other thread to be allowed to just execute the same functions until the self-test is complete.
So the lock blocks all other threads, while the thread-local storage allow us to avoid recursively taking the lock and deadlocking on the thread that is currently running the test.